### PR TITLE
WIP: Introduce support for Immutable fields in CRDs

### PR DESCRIPTION
**Pull request by enxebre**
**What type of PR is this?**
/kind api-change

**What this PR does / why we need it**:
Introduce support for Immutable fields in CRDs
https://github.com/kubernetes/enhancements/pull/1265
https://github.com/kubernetes-sigs/controller-tools/issues/339

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go [hunk 1]
```diff
@@ -995,7 +999,7 @@ func validateSimpleJSONPath(s string, fldPath *field.Path) field.ErrorList {
 	return allErrs
 }
 
-var allowedFieldsAtRootSchema = []string{"Description", "Type", "Format", "Title", "Maximum", "ExclusiveMaximum", "Minimum", "ExclusiveMinimum", "MaxLength", "MinLength", "Pattern", "MaxItems", "MinItems", "UniqueItems", "MultipleOf", "Required", "Items", "Properties", "ExternalDocs", "Example", "XPreserveUnknownFields"}
+var allowedFieldsAtRootSchema = []string{"Description", "Type", "Format", "Title", "Maximum", "ExclusiveMaximum", "Minimum", "ExclusiveMinimum", "MaxLength", "MinLength", "Pattern", "MaxItems", "MinItems", "UniqueItems", "MultipleOf", "Required", "Items", "Properties", "ExternalDocs", "Example", "XPreserveUnknownFields", "XKubernetesImmutable"}
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go]: it should be forbidden at the root.

And we need a validation that it is not possible inside of `metadata` at the root.

**Comment by apelisse** [comment]
I'd really really like if this code wasn't specific to CRDs so that we can also use it in for core types if needed.

**Comment by enxebre** [comment]
@sttts @apelisse this should be ok for first review. If this is looking good I'll proceed with integration tests

**Comment by sttts** [comment]
> I'd really really like if this code wasn't specific to CRDs so that we can also use it in for core types if needed.

The core algorithm will be generic and easily portable to any other more general place (like ssa). We can move the call to the generic registry as a follow-up. We can discuss that along the way and prototype the more generic variant. I am not concerned too much how we start as long as the desired semantics is clear.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go [hunk 2]
```diff
@@ -0,0 +1,552 @@
...
+		if _, ok := indexedList[indexKey]; !ok {
+			return false
+		}
+	}
+
+	return true
+}
+
+func isImmutableArraySetMutableValues(immutability string, old, new []interface{}) bool {
+	oldKeyCount := make(map[interface{}]int)
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go]: `make(map[interface{}]int, 0, len(old))`, saves a lot of allocations.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go [hunk 3]
```diff
@@ -0,0 +1,550 @@
...
+	structuralschema "k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+const (
+	xListTypeMap    = "map"
+	xListTypeAtomic = "atomic"
+	xListTypeSet    = "set"
+	xMapType        = "granular"
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go]: shouldn't we create constants in k8s.io/apiextensions-apiserver/pkg/apiserver/schema where Structural is defined?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go [hunk 4]
```diff
@@ -0,0 +1,550 @@
...
+		return nil
+	}
+
+	var allErrs field.ErrorList
+	switch old := old.(type) {
+	case map[string]interface{}:
+		newObj, ok := new.(map[string]interface{})
+		if !ok {
+			// e.g the CRD definition has changed
+			// TODO: what's the expectation here?
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go]: It is a mutation of `oldObj`. So if  `s.XMutability` is set, we have to reject it.

For anything further inside we don't have to check anything.

Note, that this case could happen for raw JSON with `x-kubernetes-preserve-unknown-fields:true`.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go [hunk 5]
```diff
@@ -0,0 +1,550 @@
...
+
+// Immutable validates immutable values are respected or error otherwise
+func Immutable(old interface{}, new interface{}, s *structuralschema.Structural, pth *field.Path) field.ErrorList {
+	if s == nil {
+		return nil
+	}
+
+	var allErrs field.ErrorList
+	switch old := old.(type) {
+	case map[string]interface{}:
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go]: I usually prefer to have funcs for each type, i.e. for maps and slice. And then call them each other. Easier to read than giant indentions.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go [hunk 6]
```diff
@@ -0,0 +1,550 @@
...
+		newObj, ok := new.(map[string]interface{})
+		if !ok {
+			// e.g the CRD definition has changed
+			// TODO: what's the expectation here?
+		}
+		for k, prop := range s.Properties {
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go]: there is also s.Generic.AdditionalProperties

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go [hunk 7]
```diff
@@ -0,0 +1,550 @@
...
+			// XImmutable and XImmutableKeys are mutually exclusive
+			// CRD validation should never let this be true
+			if prop.XImmutability != nil && prop.XKeyImmutability != nil {
+				allErrs = append(allErrs, field.Invalid(pth.Child(k), newObj[k], "XImmutability and XKeyImmutability are mutually exclusive"))
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go]: I think we can assume invariants here that only one is set. We check that with validation.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go [hunk 8]
```diff
@@ -0,0 +1,550 @@
...
+
+			_, foundOld := old[k]
+			_, foundNew := newObj[k]
+
+			// XImmutability
+			// validate immutable / addOnly /removeOnly against the current property regardless of the type.
+			// XImmutability/XKeyImmutability below this level is disallowed.
+			if prop.XImmutability != nil {
+				if *prop.XImmutability == xImmutabilityImmutable {
+					if (foundOld != foundNew) || !(equality.Semantic.DeepEqual(old[k], newObj[k])) {
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go]: we have to use reflect.DeepEqual everywhere, not the semantic variant.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go [hunk 9]
```diff
@@ -0,0 +1,550 @@
...
+						allErrs = append(allErrs, field.Invalid(pth.Child(k), newObj[k], "is immutable"))
+					}
+					continue
+				}
+				if *prop.XImmutability == xImmutabilityAddOnly {
+					if foundOld && !(equality.Semantic.DeepEqual(old[k], newObj[k])) {
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go]: what about: `foundOld==true`, `foundNew==false`, `old[k]=n=il` (i.e.  `null` in JSON), `newObj[k]==nil`). Then this does not trigger, but should.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go [hunk 10]
```diff
@@ -0,0 +1,550 @@
...
+			}
+
+			_, foundOld := old[k]
+			_, foundNew := newObj[k]
+
+			// XImmutability
+			// validate immutable / addOnly /removeOnly against the current property regardless of the type.
+			// XImmutability/XKeyImmutability below this level is disallowed.
+			if prop.XImmutability != nil {
+				if *prop.XImmutability == xImmutabilityImmutable {
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go]: make it a switch and drop the `continue`s.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go [hunk 11]
```diff
@@ -0,0 +1,550 @@
...
+		if _, ok := newMap[key]; ok {
+			if !equality.Semantic.DeepEqual(oldMap[key], newMap[key]) {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func derefPointerToString(pointer *string) string {
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go]: strOrEmpty

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go [hunk 12]
```diff
@@ -0,0 +1,550 @@
...
+					continue
+				}
+				continue
+			}
+
+			// Mutable array, immutable items, list type undefined / atomic
+			// The items with their respective index in the array are immutable.
+			// Hence, appending and removal at the end of the array are allowed, the change of an item or the change of the order are not.
+			if prop.Type == genericTypeArray &&
+				(derefPointerToString(prop.XListType) == "" || derefPointerToString(prop.XListType) == xListTypeAtomic) &&
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go]: shouldn't we unify `atomic` and empty string in the structural types? I don't see a reason we keep this distinction in all this logic. They are the same thing. Simular to granular and empty for maps.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go [hunk 13]
```diff
@@ -0,0 +1,550 @@
...
+					allErrs = append(allErrs, field.Invalid(pth.Child(s.Title), newArr, "is immutable"))
+					continue
+				}
+				continue
+			}
+
+			// Mutable array, immutable items, list type map
+			// The key-value pairs in the array are immutable, the set of keys is not.
+			// Hence, addition and removal of key-values pairs is allowed.
+			if prop.Type == genericTypeArray &&
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go]: all the array logic should go to the array case of the top-level switch statement. I think you are missing non-field slices now by having it here:

```
type: array
items:
  type: array
  x-kubernetes-mutability: Immutable
```

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go [hunk 14]
```diff
@@ -0,0 +1,550 @@
...
+	indexedList := make(map[string]interface{})
+	for i := range old {
+		oldItem, ok := old[i].(map[string]interface{})
+		if !ok {
+			return false
+		}
+
+		indexKey := ""
+		for _, key := range keys {
+			indexKey += fmt.Sprintf("%v", oldItem[key])
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go]: this has quadratic runtime. Always use a string buffer instead when repeatedly appending strings.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go [hunk 15]
```diff
@@ -0,0 +1,550 @@
...
+	}
+	for key := range shortestArray {
+		if !equality.Semantic.DeepEqual(old[key], new[key]) {
+			return false
+		}
+	}
+	return true
+}
+
+func isMutableArrayMapImmutableItems(keys []string, old, new []interface{}) bool {
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go]: ERRCANNOTPARSEWORD. What does it actually return?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go [hunk 16]
```diff
@@ -0,0 +1,550 @@
...
+		if !ok {
+			return false
+		}
+
+		indexKey := ""
+		for _, key := range keys {
+			indexKey += fmt.Sprintf("%v", newItem[key])
+		}
+
+		if oldItem, ok := indexedList[indexKey]; ok {
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go]: and if !ok ?

What about the converse, i.e. old had elements new doesn't ?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go [hunk 17]
```diff
@@ -0,0 +1,550 @@
...
+
+	for i := range new {
+		newItem, ok := new[i].(map[string]interface{})
+		if !ok {
+			return false
+		}
+
+		indexKey := ""
+		for _, key := range keys {
+			indexKey += fmt.Sprintf("%v", newItem[key])
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go]: are we sure this is a good encoding? I am not even sure that encoding is one-to-one, i.e. injective.

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go]: Would rather use apimachinery json marshal.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go [hunk 18]
```diff
@@ -0,0 +1,550 @@
...
+	for key := range shortestArray {
+		if !equality.Semantic.DeepEqual(old[key], new[key]) {
+			return false
+		}
+	}
+	return true
+}
+
+func isMutableArrayMapImmutableItems(keys []string, old, new []interface{}) bool {
+	indexedList := make(map[string]interface{})
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go]: this leads to many allocations. At least set a capacity from the beginning.

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/immutable_test.go [hunk 19]
```diff
@@ -0,0 +1,165 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/immutable_test.go]: 2019

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go [hunk 20]
```diff
@@ -0,0 +1,550 @@
...
+	xListTypeAtomic = "atomic"
+	xListTypeSet    = "set"
+	xMapType        = "granular"
+
+	genericTypeInteger      = "integer"
+	genericTypeNumber       = "number"
+	genericTypeString       = "string"
+	genericTypeBool         = "boolean"
+	genericTypeObject       = "object"
+	genericTypeArray        = "array"
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go]: same here: let's introduce constants for a string type alias in k8s.io/apiextensions-apiserver/pkg/apiserver/schema

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go (hunk 16 above)
Comment (enxebre) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go]: removal is allowed here so I think this is fine

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/structural.go [hunk 21]
```diff
@@ -20,6 +20,23 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 )
 
+const (
+	XListTypeMap    = "map"
+	XListTypeAtomic = "atomic"
+	XListTypeSet    = "set"
+	XMapType        = "granular"
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/structural.go]: XMapTypeAtomic

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/structural.go]: Please a string type alias to make this type safe.

Comment (enxebre) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/structural.go]: unless there're strong objections I'd prefer to do this in a different follow up PR so we also update the internal types in `Extensions struct` to the alias

**Comment by liubog2008** [comment]
how about this PR @enxebre

**Comment by enxebre** [comment]
@liubog2008 I'm hoping to revive the immutability effort and this PR. I haven't had the chance to prioritise it yet.

**Comment by apelisse** [comment]
I'm working on another API proposal that will allow to make this one much easier/better. Stay tuned!

**Comment by JoelSpeed** [comment]
@apelisse Your last comment sounds pretty interesting, do you have a link to the work you did?

**Pull request closed without merging**